File: tools/tf-types-ts/src/core/bridge-matrix.ts
/**
 * Matrix bridge — projects Matrix room events into TrustForge proof
 * events and back. The bridge does not speak the Matrix CS-API itself;
 * it only translates the typed shape so a Matrix room can be wired in
 * as a proof transport / approval channel.
 *
 * The bridge supports:
 *   - matrixEventToProofEvent: m.room.message / m.tf.* → ProofEvent
 *   - proofEventToMatrixEvent: ProofEvent → m.tf.event
 *   - signed-event canonical-JSON re-derivation so an ed25519 signature
 *     over the projected ProofEvent can be verified against the Matrix
 *     server's signed event (if the caller supplies the key).
 */

import type { ActorId, ProofLevel, Timestamp } from "../generated/_common.js";
import type { ProofEvent } from "../generated/proof-event.js";
import { BridgeFailure, type Bridge, type BridgeKind } from "./bridges.js";

export interface MatrixEvent {
  /** Matrix event identifier (e.g. `$abc123:server.example`). */
  event_id: string;
  /** Matrix room identifier (e.g. `!room:server.example`). */
  room_id: string;
  /** Matrix event type, e.g. `m.room.message` or `m.tf.proof`. */
  type: string;
  /** Matrix sender (e.g. `@alice:server.example`). */
  sender: string;
  /** Origin server timestamp (ms since epoch). */
  origin_server_ts: number;
  /** Free-form event body. */
  content: Record<string, unknown>;
  /** Optional state_key for state events. */
  state_key?: string;
  /** Optional `signatures` from the originating server. */
  signatures?: Record<string, Record<string, string>>;
}

export interface MatrixBridgeConfig {
  bridgeId: string;
  trustDomain: string;
  /** Default proof level applied to projected events when the source
   *  Matrix event doesn't carry one. */
  defaultLevel?: ProofLevel;
  /** Map a Matrix sender (e.g. `@alice:example.com`) to a TrustForge
   *  ActorId. Default: `tf:actor:human:<server>/<localpart>`. */
  senderMapper?: (sender: string) => ActorId;
  /** Optional: when the Matrix event is a `m.room.message`, this maps
   *  the textual body to a TrustForge action name. */
  messageActionMapper?: (body: string) => string | undefined;
}

export interface ProjectedMatrixEvent {
  event: ProofEvent;
  /** True when the event was a `m.room.message`. */
  fromMessage: boolean;
}

export class MatrixBridge implements Bridge {
  readonly kind: BridgeKind = "matrix";
  readonly bridgeId: string;
  readonly trustDomain: string;
  private readonly cfg: MatrixBridgeConfig;

  constructor(cfg: MatrixBridgeConfig) {
    this.bridgeId = cfg.bridgeId;
    this.trustDomain = cfg.trustDomain;
    this.cfg = cfg;
  }

  /** Project a Matrix event into a TrustForge ProofEvent. */
  matrixEventToProofEvent(matrix: MatrixEvent): ProjectedMatrixEvent {
    if (!matrix.event_id || !matrix.sender || !matrix.type) {
      throw new BridgeFailure({
        code: "invalid-input",
        message: "Matrix event missing event_id / sender / type",
      });
    }
    const actor = this.mapSender(matrix.sender);
    const timestamp = new Date(matrix.origin_server_ts).toISOString();
    const fromMessage = matrix.type === "m.room.message";
    const tfType = fromMessage
      ? this.deriveActionFromMessage(matrix)
      : matrix.type.replace(/^m\./, "matrix.");
    const event: ProofEvent = {
      event_version: "1",
      id: matrix.event_id,
      type: tfType,
      actor_id: actor,
      timestamp,
      level: this.cfg.defaultLevel ?? "L1",
      context: {
        matrix: {
          room_id: matrix.room_id,
          state_key: matrix.state_key,
          content: matrix.content,
          server_signatures: matrix.signatures,
        },
      },
      signature: {
        algorithm: "ed25519",
        signer: actor,
        // Matrix events carry server-side signatures rather than
        // TrustForge envelope signatures. Callers re-sign the projected
        // event with their own key when emitting it into a TrustForge
        // log.
        signature: "AAAA",
      },
    };
    return { event, fromMessage };
  }

  /** Project a TrustForge ProofEvent into a Matrix m.tf.event payload. */
  proofEventToMatrixEvent(event: ProofEvent, opts: { roomId: string; sender: string }): MatrixEvent {
    return {
      event_id: event.id,
      room_id: opts.roomId,
      type: "m.tf.event",
      sender: opts.sender,
      origin_server_ts: Date.parse(event.timestamp),
      content: {
        proof_event: event,
      },
    };
  }

  /** Default sender mapper: `@alice:example.com` → `tf:actor:human:example.com/alice`. */
  private mapSender(sender: string): ActorId {
    if (this.cfg.senderMapper) return this.cfg.senderMapper(sender);
    const m = /^@([^:]+):(.+)$/.exec(sender);
    if (!m) {
      throw new BridgeFailure({
        code: "invalid-input",
        message: `cannot map non-Matrix sender: ${sender}`,
      });
    }
    return `tf:actor:human:${m[2]!}/${m[1]!}`;
  }

  private deriveActionFromMessage(matrix: MatrixEvent): string {
    const body = typeof matrix.content?.body === "string" ? (matrix.content.body as string) : "";
    if (this.cfg.messageActionMapper) {
      const mapped = this.cfg.messageActionMapper(body);
      if (mapped) return mapped;
    }
    // Default: synthesize a generic matrix.message event type.
    return "matrix.message";
  }
}

export function defaultMatrixSenderMapper(sender: string): ActorId {
  const m = /^@([^:]+):(.+)$/.exec(sender);
  if (!m) throw new Error(`cannot map non-Matrix sender: ${sender}`);
  return `tf:actor:human:${m[2]!}/${m[1]!}`;
}

export function nowMatrixTs(time: Timestamp = new Date().toISOString()): number {
  return Date.parse(time);
}

File: crates/tf-types/src/bridge_matrix.rs
//! Matrix bridge — Rust mirror.

use serde::{Deserialize, Serialize};
use serde_json::{json, Value};

use crate::bridges::{Bridge, BridgeError, BridgeKind};

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct MatrixEvent {
    pub event_id: String,
    pub room_id: String,
    #[serde(rename = "type")]
    pub kind: String,
    pub sender: String,
    pub origin_server_ts: i64,
    pub content: Value,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub state_key: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub signatures: Option<Value>,
}

#[derive(Clone, Debug, Default)]
pub struct MatrixBridgeConfig {
    pub bridge_id: String,
    pub trust_domain: String,
    pub default_level: Option<String>,
}

pub struct MatrixBridge {
    cfg: MatrixBridgeConfig,
}

impl MatrixBridge {
    pub fn new(cfg: MatrixBridgeConfig) -> Self {
        MatrixBridge { cfg }
    }

    pub fn matrix_event_to_proof_event(&self, m: &MatrixEvent) -> Result<Value, BridgeError> {
        if m.event_id.is_empty() || m.sender.is_empty() || m.kind.is_empty() {
            return Err(BridgeError::InvalidInput(
                "Matrix event missing event_id / sender / type".into(),
            ));
        }
        let actor = map_sender(&m.sender)?;
        let timestamp = ms_to_iso8601(m.origin_server_ts);
        let from_message = m.kind == "m.room.message";
        let tf_type = if from_message {
            "matrix.message".to_string()
        } else if let Some(rest) = m.kind.strip_prefix("m.") {
            format!("matrix.{}", rest)
        } else {
            m.kind.clone()
        };
        Ok(json!({
            "event_version": "1",
            "id": m.event_id,
            "type": tf_type,
            "actor_id": actor,
            "timestamp": timestamp,
            "level": self.cfg.default_level.clone().unwrap_or_else(|| "L1".into()),
            "context": {
                "matrix": {
                    "room_id": m.room_id,
                    "state_key": m.state_key,
                    "content": m.content,
                    "server_signatures": m.signatures,
                }
            },
            "signature": { "algorithm": "ed25519", "signer": actor, "signature": "AAAA" }
        }))
    }
}

impl Bridge for MatrixBridge {
    fn bridge_id(&self) -> &str {
        &self.cfg.bridge_id
    }
    fn kind(&self) -> BridgeKind {
        BridgeKind::Matrix
    }
    fn trust_domain(&self) -> &str {
        &self.cfg.trust_domain
    }
}

pub fn map_sender(sender: &str) -> Result<String, BridgeError> {
    let rest = sender.strip_prefix('@').ok_or_else(|| {
        BridgeError::InvalidInput(format!("cannot map non-Matrix sender: {}", sender))
    })?;
    let colon = rest.find(':').ok_or_else(|| {
        BridgeError::InvalidInput(format!("cannot map non-Matrix sender: {}", sender))
    })?;
    let local = &rest[..colon];
    let server = &rest[colon + 1..];
    Ok(format!("tf:actor:human:{}/{}", server, local))
}

fn ms_to_iso8601(ms: i64) -> String {
    let secs = ms.div_euclid(1000);
    let (y, m, d, h, mi, s) = secs_to_ymdhms(secs);
    format!("{:04}-{:02}-{:02}T{:02}:{:02}:{:02}Z", y, m, d, h, mi, s)
}

fn secs_to_ymdhms(secs: i64) -> (i32, u32, u32, u32, u32, u32) {
    let days = secs.div_euclid(86_400);
    let time = secs.rem_euclid(86_400);
    let hour = (time / 3600) as u32;
    let minute = ((time % 3600) / 60) as u32;
    let second = (time % 60) as u32;
    let z = days + 719_468;
    let era = if z >= 0 { z } else { z - 146_096 } / 146_097;
    let doe = (z - era * 146_097) as u64;
    let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
    let y = yoe as i64 + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = (doy - (153 * mp + 2) / 5 + 1) as u32;
    let m = if mp < 10 { (mp + 3) as u32 } else { (mp - 9) as u32 };
    let year = if m <= 2 { y + 1 } else { y };
    (year as i32, m, d, hour, minute, second)
}
